py/bitbox02/bitbox02/util.py

```diff
@@ -39,3 +46,90 @@ def parse_xpub(xpub: str) -> common.XPub:
         chain_code=chain_code,
         public_key=pubkey,
     )
+
+
+class UserCache:
+    """Data structure to hold keys"""
+
+    # pylint: disable=too-few-public-methods
+    def __init__(self, raw_cache: Optional[str] = None):
+        if raw_cache is None:
+            self.app_static_privkey = None
+            self.device_static_pubkeys: List[bytes] = []
+            return
+        (privkey, pubkeys) = UserCache.deserialize(raw_cache)
+        self.app_static_privkey = privkey
+        self.device_static_pubkeys = pubkeys
+
+    def serialize(self) -> str:
+        """Serialize struct to string"""
+        pubkeys = [binascii.hexlify(x).decode("utf-8") for x in self.device_static_pubkeys]
+        privkey = None
+        if self.app_static_privkey is not None:
+            privkey = binascii.hexlify(self.app_static_privkey).decode("utf-8")
+        return json.dumps({"device_static_pubkeys": pubkeys, "app_static_privkey": privkey})
+
+    @staticmethod
+    def deserialize(raw: str) -> Tuple[Optional[bytes], List[bytes]]:
+        """Deserialize content from disk to struct"""
+        try:
+            data = json.loads(raw)
+            privkey = None
+            if data["app_static_privkey"] is not None:
+                privkey = binascii.unhexlify(data["app_static_privkey"])
+            pubkeys = [binascii.unhexlify(x) for x in data["device_static_pubkeys"]]
+            return (privkey, pubkeys)
+        except json.JSONDecodeError:
+            return (None, [])
+
+
+class NoiseConfigUserCache(bitbox_api_protocol.BitBoxNoiseConfig):
+    """A noise config that stores the keys in a file in XDG_CACHE_HOME or ~/.cache"""
+
+    def __init__(self) -> None:
+        self._cache_file_path = NoiseConfigUserCache._find_cache_file()
+        super().__init__()
+
+    @staticmethod
+    def _find_cache_file() -> Path:
+        cachedir_env = os.environ.get("XDG_CACHE_HOME", "")
+        if cachedir_env == "":
+            homedir = os.environ.get("HOME", "")
+            if homedir == "":
+                raise RuntimeError("Can't find cache dir")
+            cachedir = Path(homedir) / ".cache"
+        else:
+            cachedir = Path(cachedir_env)
+        return cachedir / "shift" / "bitbox02.dat"
```

Comment:
the whole thing should also add an appname, otherwise there is a risk that downstream projects will all use this cache util class and all use the same noise key, but there should be one per application.

Reply:
I think the opposite, I want all the apps to use the same key. In my case `send_message` and `load_firmware`. Or maybe I want all of them to use the same list of device keys...

Reply:
changed my mind. Fixed.